src/Features/CSharp/Portable/CodeRefactorings/InlineMethod/CSharpInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,149 @@
+        {
+            if (statementSyntax.IsKind(SyntaxKind.ExpressionStatement))
+            {
+                return true;
+            }
+
+            if (statementSyntax is ReturnStatementSyntax returnStatementSyntax)
+            {
+                // In this case don't provide inline.
+                // void Caller() { Callee(); }
+                // void Callee() { return; }
+                return returnStatementSyntax.Expression != null;
+            }
+
+            return false;
+        }
+
+        private static bool TryGetSingleStatementOrExpressionMethod(MethodDeclarationSyntax methodDeclarationSyntax, out ExpressionSyntax expressionSyntax)
+        {
+            var blockSyntaxNode = methodDeclarationSyntax.Body;
+            expressionSyntax = null;
+            if (blockSyntaxNode != null)
+            {
+                // 1. If it is an ordinary method with block
+                var blockStatements = blockSyntaxNode.Statements;
+                if (blockStatements.Count == 1 && CanStatementBeInlined(blockStatements[0]))
+                {
+                    expressionSyntax = GetExpressionFromStatementSyntaxNode(blockStatements[0]);
+                    return true;
+                }
+            }
+            else
+            {
+                // 2. If it is an Arrow Expression
+                var arrowExpressionNode = methodDeclarationSyntax.ExpressionBody;
+                if (arrowExpressionNode != null)
+                {
+                    expressionSyntax = arrowExpressionNode.Expression;
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private static ExpressionSyntax GetExpressionFromStatementSyntaxNode(StatementSyntax statementSyntax)
+            => statementSyntax switch
+            {
+                // Check has been done before to make sure the argument is ReturnStatementSyntax or ExpressionStatementSyntax
+                // and their expression is not null
+                ReturnStatementSyntax returnStatementSyntax => returnStatementSyntax.Expression!,
+                ExpressionStatementSyntax expressionStatementSyntax => expressionStatementSyntax.Expression,
+                _ => throw ExceptionUtilities.Unreachable
+            };
+
+        protected override bool IsSingleStatementOrExpressionMethod(MethodDeclarationSyntax calleeMethodDeclarationSyntaxNode)
+            => TryGetSingleStatementOrExpressionMethod(calleeMethodDeclarationSyntaxNode, out _);
+
+        protected override ExpressionSyntax GetInlineStatement(MethodDeclarationSyntax calleeMethodDeclarationSyntaxNode)
+        {
+            if (TryGetSingleStatementOrExpressionMethod(calleeMethodDeclarationSyntaxNode, out var expressionSyntax))
+            {
+                return expressionSyntax;
+            }
+
+            // Check has been done before to make sure it will not hit here.
+            throw ExceptionUtilities.Unreachable;
```

Comment:
woudl it perhaps be simpler to just have the caller call TryGetSingleStatementOrExpressionMethod once and have them store the expression?  They know if they ge tit that things are valid, and they dont' need to call back later to try to retrieve it.